File: android/app/src/main/java/com/openburnbar/ui/media/ScreenShareSmartZoom.kt
// Compose layout literals (dp/sp/alpha); token-per-line extraction obscures UI structure.

package com.openburnbar.ui.media

import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Rect
import androidx.compose.ui.unit.IntSize
import com.openburnbar.irohrelay.HermesRealtimeRelayFocusContext
import com.openburnbar.irohrelay.HermesRealtimeRelayFocusTargetKind
import com.openburnbar.irohrelay.HermesRealtimeRelayNormalizedPoint
import com.openburnbar.irohrelay.HermesRealtimeRelayNormalizedRect

/**
 * Mercury Smart Zoom — pure Android reducer that mirrors the iOS
 * `ScreenShareSmartZoomReducer` (see
 * `OpenBurnBarMobile/Views/Media/ScreenShareViewerView.swift`).
 *
 * The reducer is intentionally side-effect free so the JVM unit
 * suite (`./gradlew :app:testDebugUnitTest`) can exercise every
 * branch — clamping, stale context, manual override, display
 * filtering — without launching a Compose UI.
 */
enum class SmartZoomMode(val label: String) {
    OFF("Off"),
    SMART("Smart"),
    TEXT("Text"),
    WINDOW("Window"),
    CURSOR("Cursor"),
}

data class ScreenShareSmartZoomContext(
    val targetKind: HermesRealtimeRelayFocusTargetKind,
    val displayId: String?,
    val normalizedRect: HermesRealtimeRelayNormalizedRect?,
    val normalizedPoint: HermesRealtimeRelayNormalizedPoint?,
    val confidence: Double?,
    val receivedAtMillis: Long,
) {
    companion object {
        fun from(relay: HermesRealtimeRelayFocusContext, receivedAtMillis: Long = System.currentTimeMillis()): ScreenShareSmartZoomContext? {
            val kind = relay.targetKind ?: return null
            return ScreenShareSmartZoomContext(
                targetKind = kind,
                displayId = relay.displayId,
                normalizedRect = relay.normalizedRect,
                normalizedPoint = relay.normalizedPoint,
                confidence = relay.confidence,
                receivedAtMillis = receivedAtMillis,
            )
        }
    }
}

data class ScreenShareSmartZoomDecision(
    val scale: Float,
    val translation: Offset,
    val isAutoFollowing: Boolean,
) {
    companion object {
        val identity = ScreenShareSmartZoomDecision(scale = 1f, translation = Offset.Zero, isAutoFollowing = false)
    }
}

internal fun screenShareSmartZoomTargetMatches(mode: SmartZoomMode, kind: HermesRealtimeRelayFocusTargetKind): Boolean = when (mode) {
    SmartZoomMode.OFF -> false
    SmartZoomMode.SMART -> true
    SmartZoomMode.TEXT -> kind == HermesRealtimeRelayFocusTargetKind.FOCUSED_ELEMENT
    SmartZoomMode.WINDOW -> kind == HermesRealtimeRelayFocusTargetKind.FOCUSED_WINDOW
    SmartZoomMode.CURSOR -> kind == HermesRealtimeRelayFocusTargetKind.CURSOR
}

private fun screenShareSmartZoomClamp(value: Float, min: Float, max: Float): Float {
    if (value.isNaN()) return min
    return value.coerceIn(min, max)
}

object ScreenShareSmartZoomReducer {
    const val STALE_AFTER_MILLIS: Long = 1_500L
    const val MANUAL_OVERRIDE_HOLD_MILLIS: Long = 5_000L
    const val TEXT_FILL_RATIO: Float = 0.62f
    const val WINDOW_FILL_RATIO: Float = 0.86f
    const val AGENT_FILL_RATIO: Float = 0.72f
    const val MIN_SCALE: Float = 1.0f
    const val MAX_SCALE: Float = 4.0f
    val TEXT_SCALE_RANGE: ClosedFloatingPointRange<Float> = 1.4f..4.0f
    val WINDOW_SCALE_RANGE: ClosedFloatingPointRange<Float> = 1.0f..2.4f
    val AGENT_SCALE_RANGE: ClosedFloatingPointRange<Float> = 1.0f..3.0f
    const val CURSOR_ENTRY_SCALE: Float = 1.8f

    data class SmartZoomViewport(
        val viewportSize: IntSize,
        val contentRect: Rect,
        val currentScale: Float,
        val currentTranslation: Offset,
    )

    data class SmartZoomReduceInputs(
        val context: ScreenShareSmartZoomContext?,
        val mode: SmartZoomMode,
        val selectedDisplayId: String?,
        val manualOverrideUntilMillis: Long?,
        val nowMillis: Long,
    )

    fun reduce(viewport: SmartZoomViewport, inputs: SmartZoomReduceInputs): ScreenShareSmartZoomDecision {
        val idle =
            ScreenShareSmartZoomDecision(
                scale = viewport.currentScale,
                translation = viewport.currentTranslation,
                isAutoFollowing = false,
            )
        val ctx = inputs.context ?: return idle
        if (!canFollowSmartZoom(viewport, inputs, ctx)) return idle
        return reduceForTargetKind(viewport, ctx)
    }

    private fun canFollowSmartZoom(viewport: SmartZoomViewport, inputs: SmartZoomReduceInputs, ctx: ScreenShareSmartZoomContext): Boolean {
        val viewportSize = viewport.viewportSize
        val contentRect = viewport.contentRect
        return inputs.mode != SmartZoomMode.OFF &&
            viewportSize.width > 0 &&
            viewportSize.height > 0 &&
            contentRect.width > 0f &&
            contentRect.height > 0f &&
            (inputs.manualOverrideUntilMillis == null || inputs.manualOverrideUntilMillis <= inputs.nowMillis) &&
            inputs.nowMillis - ctx.receivedAtMillis <= STALE_AFTER_MILLIS &&
            (ctx.displayId == null || inputs.selectedDisplayId == null || ctx.displayId == inputs.selectedDisplayId) &&
            screenShareSmartZoomTargetMatches(inputs.mode, ctx.targetKind)
    }

    private fun reduceForTargetKind(viewport: SmartZoomViewport, ctx: ScreenShareSmartZoomContext): ScreenShareSmartZoomDecision {
        val viewportSize = viewport.viewportSize
        val contentRect = viewport.contentRect
        val currentScale = viewport.currentScale
        val idle =
            ScreenShareSmartZoomDecision(
                scale = currentScale,
                translation = viewport.currentTranslation,
                isAutoFollowing = false,
            )
        return when (ctx.targetKind) {
            HermesRealtimeRelayFocusTargetKind.FOCUSED_ELEMENT -> {
                val rect = ctx.normalizedRect
                if (rect == null) idle else fitRectDecision(rect, viewportSize, contentRect, TEXT_FILL_RATIO, TEXT_SCALE_RANGE)
            }
            HermesRealtimeRelayFocusTargetKind.FOCUSED_WINDOW -> {
                val rect = ctx.normalizedRect
                if (rect == null) idle else fitRectDecision(rect, viewportSize, contentRect, WINDOW_FILL_RATIO, WINDOW_SCALE_RANGE)
            }
            HermesRealtimeRelayFocusTargetKind.AGENT_WORKSPACE -> {
                val rect = ctx.normalizedRect
                if (rect == null) idle else fitRectDecision(rect, viewportSize, contentRect, AGENT_FILL_RATIO, AGENT_SCALE_RANGE)
            }
            HermesRealtimeRelayFocusTargetKind.CURSOR -> {
                val point = ctx.normalizedPoint
                if (point == null) {
                    idle
                } else {
                    val targetScale = if (currentScale > MIN_SCALE + 0.001f) currentScale else CURSOR_ENTRY_SCALE
                    centerPointDecision(point, viewportSize, contentRect, targetScale)
                }
            }
        }
    }

    fun fitRectDecision(
        rect: HermesRealtimeRelayNormalizedRect,
        viewportSize: IntSize,
        contentRect: Rect,
        fillRatio: Float,
        scaleRange: ClosedFloatingPointRange<Float>,
    ): ScreenShareSmartZoomDecision {
        val rectW = rect.width.toFloat().coerceAtLeast(0.0001f) * contentRect.width
        val rectH = rect.height.toFloat().coerceAtLeast(0.0001f) * contentRect.height
        val shortRectAxis = minOf(rectW, rectH)
        val shortViewportAxis = minOf(viewportSize.width.toFloat(), viewportSize.height.toFloat())
        val targetShortAxis = shortViewportAxis * fillRatio
        val rawScale = targetShortAxis / shortRectAxis.coerceAtLeast(0.0001f)
        val scale = screenShareSmartZoomClamp(rawScale, scaleRange.start, scaleRange.endInclusive)
        val centerX = contentRect.left + (rect.x + rect.width / 2.0).toFloat() * contentRect.width
        val centerY = contentRect.top + (rect.y + rect.height / 2.0).toFloat() * contentRect.height
        val translation = offsetForCenter(centerX, centerY, scale, viewportSize)
        return ScreenShareSmartZoomDecision(scale = scale, translation = translation, isAutoFollowing = true)
    }

    fun centerPointDecision(point: HermesRealtimeRelayNormalizedPoint, viewportSize: IntSize, contentRect: Rect, scale: Float): ScreenShareSmartZoomDecision {
        val clampedScale = screenShareSmartZoomClamp(scale, MIN_SCALE, MAX_SCALE)
        val centerX = contentRect.left + point.x.toFloat() * contentRect.width
        val centerY = contentRect.top + point.y.toFloat() * contentRect.height
        val translation = offsetForCenter(centerX, centerY, clampedScale, viewportSize)
        return ScreenShareSmartZoomDecision(scale = clampedScale, translation = translation, isAutoFollowing = true)
    }

    /**
     * Inverse of `applySmartZoomTransform(point)`:
     *
     *   viewX = (cx - W/2) * scale + W/2 + translationX
     *   to put viewX at W/2 ⇒ translationX = (W/2 - cx) * scale.
     */
    fun offsetForCenter(centerX: Float, centerY: Float, scale: Float, viewportSize: IntSize): Offset {
        val halfWidth = viewportSize.width / 2f
        val halfHeight = viewportSize.height / 2f
        val tx = (halfWidth - centerX) * scale
        val ty = (halfHeight - centerY) * scale
        val clampedTx = clampTranslation(tx, scale, viewportSize.width.toFloat())
        val clampedTy = clampTranslation(ty, scale, viewportSize.height.toFloat())
        return Offset(clampedTx, clampedTy)
    }

    /** Inverse-maps a viewport-space point back into the un-zoomed surface. */
    fun unzoomedPosition(position: Offset, scale: Float, translation: Offset, viewportSize: IntSize): Offset {
        val halfWidth = viewportSize.width / 2f
        val halfHeight = viewportSize.height / 2f
        if (scale <= 0.0001f) return position
        val baseX = (position.x - halfWidth - translation.x) / scale + halfWidth
        val baseY = (position.y - halfHeight - translation.y) / scale + halfHeight
        return Offset(baseX, baseY)
    }

    private fun clampTranslation(value: Float, scale: Float, dimension: Float): Float {
        if (scale <= MIN_SCALE || dimension <= 0f) return 0f
        val limit = dimension * (scale - 1f) / 2f
        return value.coerceIn(-limit, limit)
    }
}

File: OpenBurnBarMobile/Views/Media/ScreenShareSmartZoom.swift
// SmartZoom mode, focus context, reducer, and the viewport transform state.
// Extracted from ScreenShareViewerView.swift (behavior-preserving split).
import SwiftUI
@preconcurrency import AVKit
@preconcurrency import MediaPlayer
#if canImport(UIKit)
import UIKit
#endif
import OpenBurnBarMedia
import OpenBurnBarCore
import OpenBurnBarComputerUseCore

enum SmartZoomMode: String, CaseIterable, Identifiable, Sendable {
    case off
    case smart
    case text
    case window
    case cursor

    var id: String { rawValue }

    var label: String {
        switch self {
        case .off: return "Off"
        case .smart: return "Smart"
        case .text: return "Text"
        case .window: return "Window"
        case .cursor: return "Cursor"
        }
    }

    var systemImage: String {
        switch self {
        case .off: return "rectangle.dashed"
        case .smart: return "sparkles.rectangle.stack"
        case .text: return "text.cursor"
        case .window: return "rectangle.inset.filled"
        case .cursor: return "cursorarrow"
        }
    }
}

struct ScreenShareSmartZoomContext: Equatable {
    var targetKind: HermesRealtimeRelayFocusTargetKind
    var displayId: String?
    var normalizedRect: HermesRealtimeRelayNormalizedRect?
    var normalizedPoint: HermesRealtimeRelayNormalizedPoint?
    var confidence: Double?
    var receivedAt: Date

    static func from(
        _ relay: HermesRealtimeRelayFocusContext,
        receivedAt: Date = Date()
    ) -> ScreenShareSmartZoomContext? {
        guard let targetKind = relay.targetKind else { return nil }
        return ScreenShareSmartZoomContext(
            targetKind: targetKind,
            displayId: relay.displayId,
            normalizedRect: relay.normalizedRect,
            normalizedPoint: relay.normalizedPoint,
            confidence: relay.confidence,
            receivedAt: receivedAt
        )
    }
}

enum ScreenShareSmartZoomReducer {
    static let staleAfter: TimeInterval = 1.5
    static let manualOverrideHold: TimeInterval = 5.0
    static let textFillRatio: CGFloat = 0.62
    static let windowFillRatio: CGFloat = 0.86
    static let agentFillRatio: CGFloat = 0.72
    static let textScaleRange: ClosedRange<CGFloat> = 1.4...4.0
    static let windowScaleRange: ClosedRange<CGFloat> = 1.0...2.4
    static let cursorEntryScale: CGFloat = 1.8
    static let agentScaleRange: ClosedRange<CGFloat> = 1.0...3.0
    /// Scale used when a double-tap optimistically zooms toward the tapped point,
    /// before the Mac's focused-element rect arrives to refine the framing.
    static let doubleTapEntryScale: CGFloat = 2.4

    struct Decision: Equatable {
        var scale: CGFloat
        var offset: CGSize
        var isAutoFollowing: Bool
    }

    static func reduce(
        viewportSize: CGSize,
        contentRect: CGRect,
        currentState: ScreenShareViewportState,
        context: ScreenShareSmartZoomContext?,
        mode: SmartZoomMode,
        selectedDisplayId: String?,
        manualOverrideUntil: Date?,
        now: Date,
        bottomInset: CGFloat = 0
    ) -> Decision {
        let idleDecision = Decision(
            scale: currentState.scale,
            offset: currentState.offset,
            isAutoFollowing: false
        )

        guard mode != .off else { return idleDecision }
        guard viewportSize.width > 0, viewportSize.height > 0 else { return idleDecision }
        guard contentRect.width > 0, contentRect.height > 0 else { return idleDecision }
        if let manualOverrideUntil, manualOverrideUntil > now { return idleDecision }
        guard let context else { return idleDecision }
        if now.timeIntervalSince(context.receivedAt) > staleAfter { return idleDecision }
        if let target = context.displayId,
           let selected = selectedDisplayId,
           target != selected {
            return idleDecision
        }
        guard targetMatches(mode: mode, kind: context.targetKind) else { return idleDecision }

        switch context.targetKind {
        case .focusedElement:
            guard let rect = context.normalizedRect else { return idleDecision }
            return fitRectDecision(
                normalizedRect: rect,
                viewportSize: viewportSize,
                contentRect: contentRect,
                fillRatio: textFillRatio,
                scaleRange: textScaleRange,
                bottomInset: bottomInset
            )
        case .focusedWindow:
            guard let rect = context.normalizedRect else { return idleDecision }
            return fitRectDecision(
                normalizedRect: rect,
                viewportSize: viewportSize,
                contentRect: contentRect,
                fillRatio: windowFillRatio,
                scaleRange: windowScaleRange,
                bottomInset: bottomInset
            )
        case .agentWorkspace:
            guard let rect = context.normalizedRect else { return idleDecision }
            return fitRectDecision(
                normalizedRect: rect,
                viewportSize: viewportSize,
                contentRect: contentRect,
                fillRatio: agentFillRatio,
                scaleRange: agentScaleRange,
                bottomInset: bottomInset
            )
        case .cursor:
            guard let point = context.normalizedPoint else { return idleDecision }
            let targetScale: CGFloat
            if currentState.isZoomed {
                targetScale = currentState.scale
            } else {
                targetScale = min(max(cursorEntryScale, ScreenShareViewportState.minimumScale), ScreenShareViewportState.maximumScale)
            }
            return centerPointDecision(
                normalizedPoint: point,
                viewportSize: viewportSize,
                contentRect: contentRect,
                scale: targetScale,
                bottomInset: bottomInset
            )
        }
    }

    static func targetMatches(mode: SmartZoomMode, kind: HermesRealtimeRelayFocusTargetKind) -> Bool {
        switch mode {
        case .off: return false
        case .smart: return true
        case .text: return kind == .focusedElement
        case .window: return kind == .focusedWindow
        case .cursor: return kind == .cursor
        }
    }

    static func fitRectDecision(
        normalizedRect rect: HermesRealtimeRelayNormalizedRect,
        viewportSize: CGSize,
        contentRect: CGRect,
        fillRatio: CGFloat,
        scaleRange: ClosedRange<CGFloat>,
        bottomInset: CGFloat = 0
    ) -> Decision {
        let rectWidthInContent = max(0.0001, CGFloat(rect.width)) * contentRect.width
        let rectHeightInContent = max(0.0001, CGFloat(rect.height)) * contentRect.height
        let shortRectAxis = min(rectWidthInContent, rectHeightInContent)
        let shortViewportAxis = min(viewportSize.width, max(1, viewportSize.height - bottomInset))
        let targetShortAxis = shortViewportAxis * fillRatio
        let rawScale = targetShortAxis / max(shortRectAxis, 0.0001)
        let scale = clamp(rawScale, range: scaleRange)
        let centerX = contentRect.minX + CGFloat(rect.x + rect.width / 2) * contentRect.width
        let centerY = contentRect.minY + CGFloat(rect.y + rect.height / 2) * contentRect.height
        let offset = offsetForCenter(
            centerInContent: CGPoint(x: centerX, y: centerY),
            scale: scale,
            viewportSize: viewportSize,
            bottomInset: bottomInset
        )
        return Decision(scale: scale, offset: offset, isAutoFollowing: true)
    }

    static func centerPointDecision(
        normalizedPoint point: HermesRealtimeRelayNormalizedPoint,
        viewportSize: CGSize,
        contentRect: CGRect,
        scale: CGFloat,
        bottomInset: CGFloat = 0
    ) -> Decision {
        let clampedScale = ScreenShareViewportState.clampScale(scale)
        let centerX = contentRect.minX + CGFloat(point.x) * contentRect.width
        let centerY = contentRect.minY + CGFloat(point.y) * contentRect.height
        let offset = offsetForCenter(
            centerInContent: CGPoint(x: centerX, y: centerY),
            scale: clampedScale,
            viewportSize: viewportSize,
            bottomInset: bottomInset
        )
        return Decision(scale: clampedScale, offset: offset, isAutoFollowing: true)
    }

    static func normalizedCenter(of rect: HermesRealtimeRelayNormalizedRect) -> CGPoint {
        CGPoint(
            x: CGFloat(rect.x + rect.width / 2),
            y: CGFloat(rect.y + rect.height / 2)
        )
    }

    /// Offset that puts `centerInContent` at the visual center of the
    /// viewport when the content is scaled by `scale`. Inverse of
    /// `ScreenShareViewportState.viewPoint(forNormalized:in:)`:
    ///
    ///   viewX = (cx - W/2) * scale + W/2 + offsetWidth
    ///   put viewX = W/2 ⇒ offsetWidth = (W/2 - cx) * scale.
    static func offsetForCenter(
        centerInContent: CGPoint,
        scale: CGFloat,
        viewportSize: CGSize,
        bottomInset: CGFloat = 0
    ) -> CGSize {
        let halfWidth = viewportSize.width / 2
        let halfHeight = viewportSize.height / 2
        // Aim for the center of the *visible* area (above any keyboard) so the target
        // lands in view and the content uses the top of the screen.
        let lift = max(0, bottomInset) / 2
        let proposedX = (halfWidth - centerInContent.x) * scale
        let proposedY = (halfHeight - centerInContent.y) * scale - lift
        return ScreenShareViewportState.clamp(
            offset: CGSize(width: proposedX, height: proposedY),
            scale: scale,
            in: viewportSize,
            bottomInset: bottomInset
        )
    }

    static func clamp(_ value: CGFloat, range: ClosedRange<CGFloat>) -> CGFloat {
        if value.isNaN { return range.lowerBound }
        return min(max(value, range.lowerBound), range.upperBound)
    }
}

struct ScreenShareViewportState: Equatable {
    static let minimumScale: CGFloat = 1
    static let maximumScale: CGFloat = 4
    static let quickZoomScale: CGFloat = 2

    var scale: CGFloat = minimumScale
    var offset: CGSize = .zero

    var isZoomed: Bool {
        scale > Self.minimumScale + 0.001
    }

    mutating func applyMagnification(_ magnification: CGFloat, in size: CGSize) {
        scale = Self.clampScale(scale * magnification)
        offset = Self.clamp(offset: offset, scale: scale, in: size)
    }

    mutating func zoom(by multiplier: CGFloat, in size: CGSize) {
        scale = Self.clampScale(scale * multiplier)
        offset = Self.clamp(offset: offset, scale: scale, in: size)
    }

    mutating func applyTranslation(_ translation: CGSize, in size: CGSize) {
        offset = Self.clamp(offset: offset + translation, scale: scale, in: size)
    }

    mutating func toggleQuickZoom(in size: CGSize) {
        if isZoomed {
            reset()
        } else {
            scale = Self.quickZoomScale
            offset = Self.clamp(offset: .zero, scale: scale, in: size)
        }
    }

    mutating func reclamp(in size: CGSize) {
        scale = Self.clampScale(scale)
        offset = Self.clamp(offset: offset, scale: scale, in: size)
    }

    mutating func reset() {
        scale = Self.minimumScale
        offset = .zero
    }

    func preview(magnification: CGFloat, translation: CGSize, in size: CGSize) -> ScreenShareViewportState {
        let previewScale = Self.clampScale(scale * magnification)
        return ScreenShareViewportState(
            scale: previewScale,
            offset: Self.clamp(offset: offset + translation, scale: previewScale, in: size)
        )
    }

    func normalizedPoint(for point: CGPoint, in size: CGSize, contentRect: CGRect? = nil) -> (x: Double, y: Double) {
        guard size.width > 0, size.height > 0 else { return (0, 0) }
        let baseX = ((point.x - (size.width / 2) - offset.width) / scale) + (size.width / 2)
        let baseY = ((point.y - (size.height / 2) - offset.height) / scale) + (size.height / 2)
        let rect = contentRect ?? CGRect(origin: .zero, size: size)
        guard rect.width > 0, rect.height > 0 else { return (0, 0) }
        let x = min(max((baseX - rect.minX) / rect.width, 0), 1)
        let y = min(max((baseY - rect.minY) / rect.height, 0), 1)
        return (Double(x), Double(y))
    }

    func viewPoint(forNormalized normalized: CGPoint, in size: CGSize) -> CGPoint {
        guard size.width > 0, size.height > 0 else { return .zero }
        let contentX = min(max(normalized.x, 0), 1) * size.width
        let contentY = min(max(normalized.y, 0), 1) * size.height
        let viewX = ((contentX - (size.width / 2)) * scale) + (size.width / 2) + offset.width
        let viewY = ((contentY - (size.height / 2)) * scale) + (size.height / 2) + offset.height
        return CGPoint(x: min(max(viewX, 0), size.width), y: min(max(viewY, 0), size.height))
    }

    static func clampScale(_ proposed: CGFloat) -> CGFloat {
        min(max(proposed, minimumScale), maximumScale)
    }

    static func clamp(offset proposed: CGSize, scale: CGFloat, in size: CGSize, bottomInset: CGFloat = 0) -> CGSize {
        guard size.width > 0, size.height > 0 else { return .zero }

        let horizontalLimit = max(0, size.width * (scale - 1) / 2)
        let verticalLimit = max(0, size.height * (scale - 1) / 2)
        // When a keyboard covers the bottom, allow the content to ride up by half the
        // covered height so the focused region sits in the visible area instead of
        // being centered behind the keyboard (and leaving the top of the screen black).
        let lift = max(0, bottomInset) / 2

        return CGSize(
            width: min(max(proposed.width, -horizontalLimit), horizontalLimit),
            height: min(max(proposed.height, -(verticalLimit + lift)), verticalLimit)
        )
    }
}
